### PR TITLE
change GsonParsers to work with a type rather than a class

### PR DESCRIPTION
**Pull request by digitalbuddha**
@Jogan I got it to work without needing an additonal ListParser.  You'll be able to use like ```GsonParserFactory.createSourceParser(new Gson(),new TypeToken<List<Foo>>() {}.getType());```

Basically the parsers were taking a Class<T> before, now they take a `Type` so that you can use `TypeTokens` to capture the type of `List<T>` and then pass it in to the regular `GsonSourceParser`

### CODE REVIEW

**Comment by Jogan** [comment]
👍

File: middleware/src/main/java/com/nytimes/android/external/store/middleware/GsonReaderParser.java [hunk 1]
```diff
@@ -4,18 +4,19 @@
...
 import java.io.Reader;
+import java.lang.reflect.Type;
 
 import javax.inject.Inject;
 
 public class GsonReaderParser<Parsed> implements Parser<Reader, Parsed> {
 
     private final Gson gson;
-    private final Class<Parsed> parsedClass;
+    private final Type parsedClass;
```
Comment (alexio) [inline middleware/src/main/java/com/nytimes/android/external/store/middleware/GsonReaderParser.java]: rename field

File: middleware/src/main/java/com/nytimes/android/external/store/middleware/GsonStringParser.java [hunk 2]
```diff
@@ -3,15 +3,17 @@
...
 
+import java.lang.reflect.Type;
+
 import javax.inject.Inject;
 
 public class GsonStringParser<Parsed> implements Parser<String, Parsed> {
 
     private final Gson gson;
-    private final Class<Parsed> parsedClass;
+    private final Type parsedClass;
```
Comment (alexio) [inline middleware/src/main/java/com/nytimes/android/external/store/middleware/GsonStringParser.java]: rename field to `type` here as well

File: middleware/src/main/java/com/nytimes/android/external/store/middleware/GsonParserFactory.java [hunk 3]
```diff
@@ -18,67 +17,48 @@
...
     /**
      * Returns a new Parser which parses from {@link BufferedSource} to the specified type, using
      * the provided {@link Gson} instance.
      */
-    public static <T> Parser<BufferedSource, List<T>> createSourceListParser(Gson gson, TypeToken<List<T>> parsedTypeToken) {
+    public static <T> Parser<BufferedSource, T> createSourceParser(Gson gson, Type type) {
         if (gson == null) throw new IllegalArgumentException("gson cannot be null.");
-        if (parsedTypeToken == null) throw new IllegalArgumentException("type token cannot be null.");
-        return new GsonSourceListParser<>(gson, parsedTypeToken);
+        if (type == null) throw new IllegalArgumentException("parsedClass cannot be null.");
```
Comment (alexio) [inline middleware/src/main/java/com/nytimes/android/external/store/middleware/GsonParserFactory.java]: correct error string to `type cannot be null`

File: middleware/src/main/java/com/nytimes/android/external/store/middleware/GsonParserFactory.java [hunk 4]
```diff
@@ -18,67 +17,48 @@
...
 
     /**
      * Returns a new Parser which parses from {@link Reader} to the specified type, using
      * a new default configured {@link Gson} instance.
      */
-    public static <T> Parser<BufferedSource, T> createSourceParser(Class<T> parsedClass) {
-        if (parsedClass == null) throw new IllegalArgumentException("parsedClass cannot be null.");
-        return new GsonSourceParser<>(new Gson(), parsedClass);
+    public static <T> Parser<Reader, T> createReaderParser(Type type) {
+        if (type == null) throw new IllegalArgumentException("parsedClass cannot be null.");
```
Comment (alexio) [inline middleware/src/main/java/com/nytimes/android/external/store/middleware/GsonParserFactory.java]: correct error string to `type cannot be null`

File: middleware/src/main/java/com/nytimes/android/external/store/middleware/GsonParserFactory.java [hunk 5]
```diff
@@ -18,67 +17,48 @@
...
-
-    /**
-     * Returns a new Parser which parses from {@link BufferedSource} to the specified type, using
-     * the provided {@link Gson} instance.
-     */
-    public static <T> Parser<BufferedSource, T> createSourceParser(Gson gson, Class<T> parsedClass) {
-        if (gson == null) throw new IllegalArgumentException("gson cannot be null.");
-        if (parsedClass == null) throw new IllegalArgumentException("parsedClass cannot be null.");
-        return new GsonSourceParser<>(gson, parsedClass);
+        if (type == null) throw new IllegalArgumentException("parsedClass cannot be null.");
```
Comment (alexio) [inline middleware/src/main/java/com/nytimes/android/external/store/middleware/GsonParserFactory.java]: correct error string to `type cannot be null`

File: middleware/src/main/java/com/nytimes/android/external/store/middleware/GsonParserFactory.java [hunk 6]
```diff
@@ -18,67 +17,48 @@
...
 
     /**
      * Returns a new Parser which parses from {@link Reader} to the specified type, using
      * a new default configured {@link Gson} instance.
      */
-    public static <T> Parser<BufferedSource, List<T>> createSourceListParser(TypeToken<List<T>> parsedTypeToken) {
-        if (parsedTypeToken == null) throw new IllegalArgumentException("type token cannot be null.");
-        return new GsonSourceListParser<>(new Gson(), parsedTypeToken);
+    public static <T> Parser<BufferedSource, T> createSourceParser(Type type) {
+        if (type == null) throw new IllegalArgumentException("parsedClass cannot be null.");
```
Comment (alexio) [inline middleware/src/main/java/com/nytimes/android/external/store/middleware/GsonParserFactory.java]: correct error string to `type cannot be null`

File: middleware/src/main/java/com/nytimes/android/external/store/middleware/GsonParserFactory.java [hunk 7]
```diff
@@ -18,67 +17,48 @@
...
     /**
      * Returns a new Parser which parses from a String to the specified type, using
      * the provided {@link Gson} instance.
      */
-    public static <T> Parser<String, T> createStringParser(Gson gson, Class<T> parsedClass) {
+    public static <T> Parser<String, T> createStringParser(Gson gson,Type type) {
         if (gson == null) throw new IllegalArgumentException("gson cannot be null.");
-        if (parsedClass == null) throw new IllegalArgumentException("parsedClass cannot be null.");
-        return new GsonStringParser<>(gson, parsedClass);
+        if (type == null) throw new IllegalArgumentException("parsedClass cannot be null.");
```
Comment (alexio) [inline middleware/src/main/java/com/nytimes/android/external/store/middleware/GsonParserFactory.java]: correct error string to `type cannot be null`

**Pull request merged**